adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/MqttContext.java

```diff
@@ -152,11 +156,22 @@ public Device authenticatedDevice() {
 
     /**
      * Gets the content type of the message payload.
+     * <p>
+     * The type is the value set via {@link #setContentType(String)}.
+     * Otherwise, the type is determined from the message topic's property
+     * bag, if if contains a content type.
+     * Otherwise, the {@linkplain MessageHelper#CONTENT_TYPE_OCTET_STREAM default
+     * content type} is used.
      *
-     * @return The type or {@code null} if the content type is unknown.
+     * @return The type of the message payload.
      */
     public String contentType() {
-        return contentType;
+        if (contentType != null) {
+            return contentType;
+        }
+        return Optional.ofNullable(propertyBag)
+                .flatMap(bag -> Optional.ofNullable((String) bag.getProperty(HttpHeaders.CONTENT_TYPE.toString())))
+                .orElse(MessageHelper.CONTENT_TYPE_OCTET_STREAM);
```

Comment:
(Redundant String cast.)
The extra propertyBag check looks superfluous here since all `MqttContext` fields except `timer` are effectively final. (And there is a different content-type constant being used compared to the constructor code.)

Reply:
Yes, I stumbled across that as well during the refactoring. I will remove the *setContentType* method and make the contentType field final, initializing it from the property bag during creation ...